File: swift_wrapper/PureParser.swift
//
//  PureParser.swift
//  PureParser
//
//  Copyright © 2019 JivoSite Inc. All rights reserved.
//  <For detailed info about how this parser works, please refer to README.md file>
//

import Foundation

#if canImport(PureParserC)
import PureParserC
#endif

public protocol IPureParser: class {
    func assign(variable name: String, value: String?)
    func activate(alias: String, _ rule: Bool)
    func execute(_ formula: String, collapseSpaces: Bool, resetOnFinish: Bool) -> String
}

@objc(PureParser) public final class PureParser: NSObject, IPureParser {
    private var config = pure_config_t()
    private var parser = pure_parser_t()
    
    public override init() {
        super.init()
        pure_config_set_default(&config)
        pure_parser_init(&parser, &config)
    }
    
    deinit {
        pure_parser_destroy(&parser)
    }
    
    @objc(assignVariable:value:) public func assign(variable name: String, value: String?) {
        if let value = value {
            pure_parser_assign_var(&parser, name, value)
        }
        else {
            pure_parser_discard_var(&parser, name)
        }
    }
    
    @objc(activateAlias:rule:) public func activate(alias: String, _ rule: Bool) {
        if rule {
            pure_parser_enable_alias(&parser, alias)
        }
        else {
            pure_parser_discard_alias(&parser, alias)
        }
    }
    
    @objc(execute:collapseSpaces:resetOnFinish:) public func execute(_ formula: String, collapseSpaces: Bool, resetOnFinish: Bool) -> String {
        var result: UnsafeMutablePointer<Int8>?
        var result_len = UInt(0)
        pure_parser_execute(&parser, formula, collapseSpaces, resetOnFinish, &result, &result_len)
        
        if let result = result, result_len > 0 {
            return String(bytesNoCopy: result, length: Int(result_len), encoding: .utf8, freeWhenDone: true) ?? String()
        }
        else {
            return String()
        }
    }
}

File: c_wrapper/pure_parser.cpp
//
//  pure_parser.cpp
//  PureParser
//
//  Copyright © 2019 JivoSite Inc. All rights reserved.
//  <For detailed info about how this parser works, please refer to README.md file>
//

#include "pure_parser.h"
#include "../cpp_src/PureParser.hpp"
#include <regex>
#include <iostream>
#include <optional>

static inline PureParser* get_impl(pure_parser_t *parser);

void pure_config_set_default(pure_config_t *config) {
    config->element_token = kPureParserDefaultElementToken;
    config->block_opener_token = kPureParserDefaultBlockOpenerToken;
    config->block_closer_token = kPureParserDefaultBlockCloserToken;
    config->separator_token = kPureParserDefaultSeparatorToken;
    config->alias_token = kPureParserDefaultAliasToken;
}

void pure_parser_init(pure_parser_t *parser, const pure_config_t *config) {
    PureConfig impl_config {
        .element_token = config->element_token,
        .block_opener_token = config->block_opener_token,
        .block_closer_token = config->block_closer_token,
        .separator_token = config->separator_token,
        .alias_token = config->alias_token
    };

    parser->impl = new PureParser(impl_config);
}

void pure_parser_reset(pure_parser_t *parser) {
    get_impl(parser)->reset();
}

void pure_parser_destroy(pure_parser_t *parser) {
    delete get_impl(parser);
}

void pure_parser_assign_var(pure_parser_t *parser, const char *name, const char *value) {
    get_impl(parser)->assignVariable(name, value);
}

void pure_parser_discard_var(pure_parser_t *parser, const char *name) {
    get_impl(parser)->discardVariable(name);
}

void pure_parser_enable_alias(pure_parser_t *parser, const char *name) {
    get_impl(parser)->enableAlias(name);
}

void pure_parser_discard_alias(pure_parser_t *parser, const char *name) {
    get_impl(parser)->disableAlias(name);
}

void pure_parser_execute(pure_parser_t *parser, const char *formula, bool collapse_spaces, bool reset_on_finish, char **output, unsigned long *output_len) {
    const std::string result = get_impl(parser)->execute(formula, collapse_spaces, reset_on_finish);
    const size_t result_len = result.size();
    
    if (output_len) {
        *output_len = result_len;
    }
    
    if (result_len > 0) {
        *output = (char *) malloc(result_len + 1);
        memcpy(*output, result.data(), result_len);
        (*output)[result_len] = 0x00;
    }
    else {
        *output = NULL;
    }
}

static inline PureParser* get_impl(pure_parser_t *parser) {
    return (PureParser *) parser->impl;
}
